sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala

```diff
@@ -123,7 +123,11 @@ object JavaTypeInference {
         val beanInfo = Introspector.getBeanInfo(typeToken.getRawType)
         val properties = beanInfo.getPropertyDescriptors.filterNot(_.getName == "class")
```

Comment:
> It seems like behavior should be the same in this case, in this same file. I think you could also improve that error handling.

The check he pointed out is related only with `Encoders.bean(...)` API as it seems only used in `deserializerFor/serializerFor` via [ExpressionEncoder.scala#L85](https://github.com/apache/spark/blob/52d4f61941b0ac8dcdca04aac042a492141c2068/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala#L85), which seems always executed after the codes I proposed to fix here.

The below codes just do not allow a bean with empty property after the properties without getters are filtered in [JavaTypeInference.scala#L137](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala#L137) and it checks if they are empty in [JavaTypeInference.scala#L421](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala#L421).

```java
spark.createDataset(data, Encoders.bean(EmptyBean.class));
```

it throws an exception as below:

```java
java.lang.UnsupportedOperationException: Cannot infer type for class EmptyBean because it is not bean-compliant
	at org.apache.spark.sql.catalyst.JavaTypeInference$.org$apache$spark$sql$catalyst$JavaTypeInference$$serializerFor(JavaTypeInference.scala:437)
	at org.apache.spark.sql.catalyst.JavaTypeInference$.serializerFor(JavaTypeInference.scala:342)
```

If we do not use `Encoders.bean(...)`, then it allows empty schema as below:

```java
spark.createDataFrame(data, EmptyBean.class);
```

For the sake of consistency, I guess we might have to consider removing the checks there (not here in this PR) and allowing this empty case because this seems possible in Scala API too as below:


```scala
import spark.implicits._

case class A()
spark.createDataset(Seq(A()))
spark.createDataFrame(Seq(A()))
```

> Is there anywhere else in the code that checks the getter? is the error message consistent with that?

It seems they are all [being filtered out](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala#L137) in the `Encoders.bean(...)` code path. It seems not being applied in this code path due to https://github.com/apache/spark/pull/17013#issuecomment-281291907.